types/validator_set.go

```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+			return fmt.Errorf("failed to find validator %X to remove", address)
+		}
+	}
+	return nil
+}
+
+// Removes the validators specified in 'deletes' from validator set 'vals'.
+// Should not fail as verification has been done before.
+func (vals *ValidatorSet) applyRemovals(deletes []*Validator) {
+
+	for _, valUpdate := range deletes {
+		address := valUpdate.Address
+		_, removed := vals.Remove(address)
+		if !removed {
+			// Should never happen
+			panic(fmt.Sprintf("failed to remove validator %X", address))
+		}
+	}
+}
+
+// Removes the validators specified in 'deletes' from validator set 'vals'.
+// 'deletes' must be sorted.
+// Should not fail as verification has been done before.
+func (vals *ValidatorSet) applyRemovals2(deletes []*Validator) {
+
+	existing := make([]*Validator, len(vals.Validators))
+	copy(existing, vals.Validators)
+
+	merged := make([]*Validator, len(existing)-len(deletes))
+	i := 0
+
+	for len(existing) > 0 && len(deletes) > 0 {
+		if bytes.Equal(existing[0].Address, deletes[0].Address) {
+			deletes = deletes[1:]
+		} else {
+			merged[i] = existing[0]
+			i++
+		}
+		existing = existing[1:]
+	}
+	for j := 0; j < len(existing); j++ {
+		merged[i] = existing[j]
+		i++
+	}
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
+}
+
+// UpdateWithChangeSet attempts to update the validator set with 'changes'
+// It performs the following steps:
+// - validates the changes making sure there are no duplicates and splits them in updates and deletes
+// - verifies that applying the changes will not result in errors
+// - computes the priorities of new validators against the final set
+// - applies the updates against the validator set
+// - applies the removals against the validator set
+// - performs scaling and centering of priority values
+// If error is detected during verification steps it is returned and the validator set
+// is not changed.
+func (vals *ValidatorSet) UpdateWithChangeSet(changes []*Validator) error {
+	return vals.updateWithChangeSet(changes, true)
+}
+
+// main function used by UpdateWithChangeSet() and NewValidatorSet()
+// If 'allowDeletes' is false then delete operations are not allowed and must be reported if
+// present in 'changes'
+func (vals *ValidatorSet) updateWithChangeSet(changes []*Validator, allowDeletes bool) error {
+
+	if len(changes) <= 0 {
+		return nil
+	}
+
+	// Check for duplicates within changes, split in 'updates' and 'deletes' lists (sorted)
+	updates, deletes, err := processChanges(changes)
+	if err != nil {
+		return err
+	}
+
+	if !allowDeletes && len(deletes) != 0 {
+		err = fmt.Errorf("cannot process validators with voting power 0: %v", deletes)
+		return err
+	}
+
+	// Verify that applying the 'deletes' against 'vals' will not result in error.
+	if err = verifyRemovals(deletes, vals); err != nil {
+		return err
+	}
+
+	// Verify that applying the 'updates' against 'vals' will not result in error.
+	updatedTotalVotingPower, err := verifyUpdates(updates, vals)
```

Comment:
don't we need to include changes from deletes in the updatedTotalVotingPower ?

Reply:
Ouch!!! Totally missed this! Thanks, thanks!!!!

Reply:
I will write a test for this soon

Reply:
Hmm, actually it was correct before. Since we start with totalVotingPower and add the deltas for changes and power for adds.
But this proves that I am definitely missing explanations...and the test.

Reply:
But we don't subtract the deletes?

Reply:
It seems I also misunderstand something. Shouldn't priority of new validator be equal to -1.25*totalVotingPower of the validator set before we apply any changes?

Reply:
No we don't, it's on purpose. So the idea is to forget the deletes initially. And determine how the set would look like without them. This way the relative priority of the existing validators is kept and we make sure we add the new validators with smallest priority (-TVP' of this imaginary set). This keeps things fair across this set.
Then we delete the required validators, scale and center. When we do that we do not change the relative priority of remaining validators and we make sure the prio window stays centered and within the 2*TVP
We discussed this with @cwgoes and @ismail in https://github.com/tendermint/tendermint/issues/3181.

Reply:
@milosevic - yes it used to be like that and I documented it thinking it was intentional. However, when I brought this up to @cwgoes and @Liamsi they mentioned that it is in fact a bug. So I changed it here...doc still needs to be updated.